File: weight_tracker/体重记录.js
/**
 * 体重记录插件 v2.0.0
 * 基于autMan实际API结构重写
 * 功能: 体重记录、趋势分析、目标管理
 * 
 * 使用说明:
 * - 发送「体重 [数值]」→ 记录当前体重 (如: 体重 65.5)
 * - 发送「体重记录 [日期] [数值]」→ 补录历史数据
 * - 发送「体重记录」→ 查看最近记录和趋势
 * - 发送「体重详细记录」→ 查看带编号的完整记录
 * - 发送「体重统计」→ 查看统计信息
 * - 发送「设置目标体重 [数值]」→ 设定目标体重
 * - 发送「目标进度」→ 查看目标进度
 * - 发送「删除体重记录 [编号]」→ 删除指定记录
 * - 发送「修改体重记录 [编号] [新数值]」→ 修改指定记录
 * - 发送「清空体重记录」→ 清空所有记录
 * - 发送「体重帮助」→ 显示帮助
 * 
 * 交互说明:
 * - 关键操作需要回复 Y/y 确认
 * - 回复 Q/q 或 N/n 取消操作
 * - 30秒无操作自动退出
 * 
 * 更新历史:
 * v2.0.0 - 重构为 input() 模式，彻底解决指令冲突
 * v1.1.6 - 恢复独立Q监听，添加防冲突协同逻辑
 */

// [disable:false]
// [rule: (.*体重.*|.*目标.*)]
// [admin: false] 
// [service: 88489948]
// [price: 0.00]
// [version: v2.0.0]
// [update: 重构为 input() 模式]

// 定义存储桶名称
(function () {
    // 定义存储桶名称 (放入 IIFE 防止与其他插件冲突)
    const BUCKET_NAME = "weight_tracker";

    /**
     * 获取当前日期字符串 (YYYY-MM-DD)
     */
    function getCurrentDate() {
        const now = new Date();
        const year = now.getFullYear();
        const month = String(now.getMonth() + 1).padStart(2, '0');
        const day = String(now.getDate()).padStart(2, '0');
        return `${year}-${month}-${day}`;
    }

    /**
     * 发送消息
     */
    function sendMessage(text) {
        try {
            if (typeof Sender !== 'undefined' && Sender && typeof Sender.reply === 'function') {
                Sender.reply(text);
                return;
            }
            if (this && this.Sender && typeof this.Sender.reply === 'function') {
                this.Sender.reply(text);
                return;
            }
            if (typeof reply === 'function') {
                reply(text);
                return;
            }
            if (typeof sendText === 'function') {
                sendText(text);
                return;
            }
            console.log("[发送消息]", text);
        } catch (error) {
            console.error("[发送消息失败]", error);
        }
    }

    /**
     * 获取消息内容
     */
    function getMessageContent() {
        if (this && this.Sender && this.Sender.sender && this.Sender.sender.message) {
            return this.Sender.sender.message.text || "";
        }
        if (this && this.Sender && this.Sender.sender && this.Sender.sender.baseSender) {
            return this.Sender.sender.baseSender.content || "";
        }
        if (this && this.Sender && this.Sender.sender) {
            return this.Sender.sender.content || "";
        }
        return "";
    }

    /**
     * 获取用户ID
     */
    function getUserID() {
        if (this && this.Sender && this.Sender.sender && this.Sender.sender.message) {
            const msg = this.Sender.sender.message;
            if (msg.sender && msg.sender.iD) return String(msg.sender.iD);
            if (msg.chat && msg.chat.iD) return String(msg.chat.iD);
        }
        return "unknown";
    }

    /**
     * 获取用户名
     */
    function getUserName() {
        if (this && this.Sender && this.Sender.sender && this.Sender.sender.message) {
            const sender = this.Sender.sender.message.sender;
            if (sender) {
                const name = `${sender.firstName || ''} ${sender.lastName || ''}`.trim();
                return name || sender.username || "用户";
            }
        }
        return "用户";
    }

    // 封装 Bucket 操作
    function safeBucketGet(key) {
        if (typeof bucketGet === 'function') return bucketGet(BUCKET_NAME, key);
        return "";
    }

    function safeBucketSet(key, value) {
        if (typeof bucketSet === 'function') bucketSet(BUCKET_NAME, key, value);
    }

    function safeBucketDel(key) {
        if (typeof bucketDel === 'function') bucketDel(BUCKET_NAME, key);
    }

    // 等待用户确认 (封装 input)
    function waitForConfirm(promptText) {
        sendMessage(promptText);
        // 等待30秒
        const userReply = input(30000);
        if (!userReply) return false; // 超时

        const trimmed = userReply.trim().toLowerCase();
        if (trimmed === 'y') return true;
        return false;
    }

    // 获取数据辅助函数
    function getData() {
        const userID = getUserID();
        const STORAGE_KEY = `user_${userID}`;
        const existingData = safeBucketGet(STORAGE_KEY);
        let data = { records: [], target: null };
        if (existingData) {
            try { data = JSON.parse(existingData); } catch (e) { }
            if (!data.records) data.records = []; // 修复旧数据结构
        }
        return data;
    }

    // 保存数据辅助函数
    function saveData(data) {
        const userID = getUserID();
        const STORAGE_KEY = `user_${userID}`;
        safeBucketSet(STORAGE_KEY, JSON.stringify(data));
    }

    // 处理记录体重
    function handleRecord(content) {
        // 解析指令
        // 格式1: 体重 60
        // 格式2: 体重记录 2023-01-01 60
        let weight = null;
        let date = getCurrentDate();

        // 简单的正则解析
        const numMatch = content.match(/(\d+(\.\d+)?)/g);
        const dateMatch = content.match(/\d{4}-\d{2}-\d{2}/);

        if (dateMatch) date = dateMatch[0];

        // 尝试从数字中找体重 (如果找到多个，且有一个匹配了日期格式，则排除日期的数字部分... 比较复杂，这里简化处理)
        // 假设用户输入: "体重记录 2026-01-01 65.5" -> nums: 2026, 01, 01, 65.5
        // 简单的策略：取最后一个数字作为体重，如果它看起来像合理的体重 (0-500)
        if (numMatch) {
            const candidate = parseFloat(numMatch[numMatch.length - 1]);
            if (candidate > 0 && candidate < 500) {
                weight = candidate;
            }
        }

        if (!weight) {
            sendMessage("❌ 请输入有效的体重数值 (如: 体重 65.5)");
            return;
        }

        const userName = getUserName();
        const confirmed = waitForConfirm(`📝 准备记录 ${userName} 在 ${date} 的体重: ${weight}kg\n\n确认记录请回复 Y, 取消请回复 Q`);

        if (confirmed) {
            const data = getData();
            // 检查当天是否已有
            const idx = data.records.findIndex(r => r.date === date);
            let msg = "";

            if (idx >= 0) {
                const old = data.records[idx].weight;
                data.records[idx].weight = weight;
                data.records[idx].timestamp = new Date().getTime();
                msg = `✅ 已更新 ${date} 的记录: ${old}kg → ${weight}kg`;
            } else {
                data.records.push({
                    date: date,
                    weight: weight,
                    timestamp: new Date().getTime()
                });
                msg = `✅ 已记录 ${date} 的体重: ${weight}kg`;
            }

            // 排序
            data.records.sort((a, b) => a.date.localeCompare(b.date));
            saveData(data);

            // 目标检查
            if (data.target) {
                const diff = weight - data.target;
                if (Math.abs(diff) < 0.1) msg += `\n🎉 达成目标!`;
                else if (diff > 0) msg += `\n📊 距目标还差 ${diff.toFixed(1)}kg`;
                else msg += `\n📊 距目标还差 ${Math.abs(diff).toFixed(1)}kg`;
            }
            sendMessage(msg);
        } else {
            sendMessage("已取消");
        }
    }

    // 处理设置目标
    function handleSetTarget(content) {
        const match = content.match(/(\d+(\.\d+)?)/);
        if (!match) {
            sendMessage("❌ 请输入有效数值");
            return;
        }
        const target = parseFloat(match[0]);

        if (waitForConfirm(`🎯 确认设置目标体重为: ${target}kg?`)) {
            const data = getData();
            data.target = target;
            saveData(data);
            sendMessage(`✅ 目标已设为 ${target}kg`);
        } else {
            sendMessage("已取消");
        }
    }

    // 处理目标进度
    function showProgress() {
        const data = getData();
        if (!data.target) {
            sendMessage("❌ 未设置目标");
            return;
        }
        if (data.records.length === 0) {
            sendMessage("📋 暂无记录");
            return;
        }

        // 取最新一条（最后一条，因为已排序）
        const latest = data.records[data.records.length - 1];
        const diff = latest.weight - data.target;

        let msg = `🎯 目标进度\n当前: ${latest.weight}kg\n目标: ${data.target}kg\n`;
        if (Math.abs(diff) < 0.1) msg += "🎉 已达标!";
        else if (diff > 0) msg += `📊 需减重 ${diff.toFixed(1)}kg`;
        else msg += `📊 需增重 ${Math.abs(diff).toFixed(1)}kg`;

        sendMessage(msg);
    }

    // 列表显示
    function showRecords(detailed) {
        const data = getData();
        if (data.records.length === 0) {
            sendMessage("📋 暂无记录");
            return;
        }
        // 反序显示（最新在前）
        const list = data.records.slice().reverse();

        if (detailed) {
            let msg = `📋 详细记录 (共${list.length}条)\n━━━━━━━━━━\n`;
            list.forEach((r, i) => {
                msg += `[${i + 1}] ${r.date}  ${r.weight}kg\n`;
            });
            msg += "\n💡 (30秒内) 发送数字可快速删除\n(回复 Q 退出)";
            sendMessage(msg);

            // 等待快速删除
            const reply = input(30000);
            if (reply && /^\d+$/.test(reply)) {
                handleDeleteIndex(parseInt(reply), list); // 注意 list 是反序的
            } else if (reply && /^[Qq]$/i.test(reply)) {
                sendMessage("✅ 已退出");
            }
        } else {
            // 简略（时间轴）
            const recent = list.slice(0, 7);
            let msg = `📊 最近记录\n`;
            recent.forEach(r => {
                msg += `🗓️ ${r.date}  ${r.weight}kg\n`;
            });
            if (list.length > 7) msg += `... 等共${list.length}条`;
            sendMessage(msg);
        }
    }

    // 删除指定索引（基于反序后的 index 1-based）
    function handleDeleteIndex(index, reversedList) {
        if (index < 1 || index > reversedList.length) {
            sendMessage("❌ 编号无效");
            return;
        }
        const target = reversedList[index - 1];
        if (waitForConfirm(`🗑️ 确认删除 [${index}]: ${target.date} ${target.weight}kg ?`)) {
            const data = getData();
            // 在原数据中找到对应日期删除 (日期是唯一的)
            const realIdx = data.records.findIndex(r => r.date === target.date);
            if (realIdx >= 0) {
                data.records.splice(realIdx, 1);
                saveData(data);
                sendMessage("✅ 删除成功");
            } else {
                sendMessage("❌ 数据同步错误");
            }
        } else {
            sendMessage("已取消");
        }
    }

    // 处理删除指令
    function handleDeleteCommand(content) {
        const match = content.match(/(\d+)/);
        if (match) {
            // 用户直接指定了编号，但这里的编号通常是指详细列表里的顺序
            // 为了安全，建议先看列表。或者这里直接获取列表并删除
            const data = getData();
            const list = data.records.slice().reverse();
            handleDeleteIndex(parseInt(match[1]), list);
        } else {
            showRecords(true);
        }
    }

    // 处理修改指令
    function handleModify(content) {
        // 格式: 修改体重记录 1 60
        const parts = content.split(/\s+/);
        if (parts.length < 3) {
            sendMessage("❌ 格式错误，例: 修改体重记录 1 60.5");
            return;
        }
        const index = parseInt(parts[1]);
        const newWeight = parseFloat(parts[2]);

        const data = getData();
        const list = data.records.slice().reverse();

        if (index < 1 || index > list.length) {
            sendMessage("❌ 编号无效");
            return;
        }

        const target = list[index - 1];
        if (waitForConfirm(`📝 将 [${index}] ${target.date} 从 ${target.weight}kg 修改为 ${newWeight}kg ?`)) {
            const realIdx = data.records.findIndex(r => r.date === target.date);
            if (realIdx >= 0) {
                data.records[realIdx].weight = newWeight;
                data.records[realIdx].timestamp = new Date().getTime();
                saveData(data);
                sendMessage("✅ 修改成功");
            }
        } else {
            sendMessage("已取消");
        }
    }

    function handleClear() {
        if (waitForConfirm("⚠️ 确认清空所有数据？(不可恢复)")) {
            safeBucketDel(`user_${getUserID()}`);
            sendMessage("🗑️ 已清空");
        } else {
            sendMessage("已取消");
        }
    }

    function showStatistics() {
        const data = getData();
        if (data.records.length === 0) { sendMessage("暂无记录"); return; }
        const weights = data.records.map(r => r.weight);
        const max = Math.max(...weights);
        const min = Math.min(...weights);
        const avg = (weights.reduce((a, b) => a + b, 0) / weights.length).toFixed(1);

        sendMessage(`📊 统计信息\n最高: ${max}kg\n最低: ${min}kg\n平均: ${avg}kg\n记录数: ${weights.length}`);
    }

    function showHelp() {
        sendMessage(`📖 体重记录 v2.0.0
体重 60 → 记录
体重记录 → 查看
体重详细记录 → 管理
删除体重记录 [编号]
修改体重记录 [编号] [数值]
体重统计
设置目标体重 [数值]
目标进度
清空体重记录`);
    }

    function main() {
        const content = getMessageContent().trim();

        // 简单路由
        if (content.startsWith("设置目标体重")) {
            handleSetTarget(content);
        } else if (content === "目标进度") {
            showProgress();
        } else if (content === "体重统计") {
            showStatistics();
        } else if (content.startsWith("修改体重记录")) {
            handleModify(content);
        } else if (content.startsWith("删除体重记录")) {
            handleDeleteCommand(content);
        } else if (content.startsWith("清空体重记录")) {
            handleClear();
        } else if (content === "体重详细记录") {
            showRecords(true);
        } else if (content === "体重记录") {
            showRecords(false);
        } else if (content === "体重帮助") {
            showHelp();
        } else if (content.startsWith("体重 ") || content.startsWith("体重记录 ")) {
            // "体重 60" 或 "体重记录 2023-01-01 60"
            handleRecord(content);
        }
    }

    main();
})();

File: weight_tracker/体重记录.py
# [disable:false]
# [rule: ^体重(.*)$]
# [admin: false]
# [price: 0.00]
# [version: v2.0.0]

"""
autMan 插件 - 体重记录

功能: 体重记录、趋势分析、目标管理
作者: AI Assistant
版本: v2.0.0
日期: 2026-01-12

使用说明:
- 体重 [数值]: 记录当前体重 (如: 体重 65.5)
- 体重记录: 查看最近记录
- 体重统计: 查看统计信息
- 设置目标体重 [数值]: 设定目标体重
- 目标进度: 查看目标进度
- 体重帮助: 显示帮助
"""

import middleware
import time
import json
import re
from datetime import datetime

# 配置常量
BUCKET_NAME = "weight_tracker"
VERSION = "v2.0.0"
INPUT_TIMEOUT = 60000  # 60秒超时


class WeightPlugin:
    def __init__(self):
        """初始化插件"""
        sender_id = middleware.getSenderID()
        self.sender = middleware.Sender(sender_id)
        self.user_id = self.sender.getUserID()
        self.username = self.user_id
        self.message = self.sender.getMessage().strip()
    
    def get_user_confirmation(self, prompt):
        """获取用户确认"""
        self.sender.reply(f"{prompt}\n\n请输入：\n  y - 确认\n  n - 取消\n  q - 退出")
        user_input = self.sender.listen(INPUT_TIMEOUT)
        
        if user_input is None:
            self.sender.reply("⏱️ 操作超时，已自动取消")
            return None
        
        return user_input.strip().lower()
    
    def get_data(self):
        """获取用户数据"""
        try:
            data = middleware.bucketGet(BUCKET_NAME, self.user_id)
            if not data or data == '':
                return {'records': [], 'target': None}
            return json.loads(data)
        except:
            return {'records': [], 'target': None}
    
    def save_data(self, data):
        """保存用户数据"""
        try:
            json_data = json.dumps(data, ensure_ascii=False)
            middleware.bucketSet(BUCKET_NAME, self.user_id, json_data)
        except Exception as e:
            self.sender.reply(f"❌ 保存失败：{e}")
    
    def record_weight(self, weight_str):
        """记录体重"""
        try:
            weight = float(weight_str)
            if weight <= 0 or weight > 500:
                self.sender.reply("❌ 体重数值无效,请输入0-500之间的数字")
                return
        except ValueError:
            self.sender.reply("❌ 体重数值无效,请输入0-500之间的数字")
            return
        
        # 请求确认
        date = datetime.now().strftime("%Y-%m-%d")
        prompt = f"📝 确认要记录 {self.username} 在 {date} 的体重: {weight}kg 吗？"
        confirmation = self.get_user_confirmation(prompt)
        
        if not confirmation:
            return
        
        if confirmation == "q":
            self.sender.reply("👋 已退出记录流程")
            return
        
        if confirmation == "n":
            self.sender.reply("❌ 已取消记录")
            return
        
        if confirmation == "y":
            # 执行记录
            data = self.get_data()
            
            # 检查当天是否已有记录
            existing_index = -1
            for i, record in enumerate(data['records']):
                if record['date'] == date:
                    existing_index = i
                    break
            
            if existing_index >= 0:
                # 更新记录
                old_weight = data['records'][existing_index]['weight']
                data['records'][existing_index]['weight'] = weight
                data['records'][existing_index]['timestamp'] = int(time.time() * 1000)
                
                self.save_data(data)
                
                diff = weight - old_weight
                diff_str = f"+{diff:.1f}" if diff > 0 else f"{diff:.1f}"
                message = f"✅ 已更新 {date} 的体重记录:\n"
                message += f"{old_weight}kg → {weight}kg ({diff_str}kg)\n\n"
                message += f"当前共有 {len(data['records'])} 条记录"
                self.sender.reply(message)
            else:
                # 添加新记录
                data['records'].append({
                    'date': date,
                    'weight': weight,
                    'timestamp': int(time.time() * 1000)
                })
                
                data['records'].sort(key=lambda x: x['date'])
                self.save_data(data)
                
                message = f"✅ 已记录 {date} 的体重: {weight}kg\n\n"
                message += f"当前共有 {len(data['records'])} 条记录"
                
                # 如果设置了目标,显示进度
                if data.get('target'):
                    diff = weight - data['target']
                    if abs(diff) < 0.1:
                        message += f"\n\n🎉 恭喜!已达成目标体重 {data['target']}kg!"
                    elif diff > 0:
                        message += f"\n\n📊 距离目标体重还差: {diff:.1f}kg (需减重)"
                    else:
                        message += f"\n\n📊 距离目标体重还差: {abs(diff):.1f}kg (需增重)"
                
                self.sender.reply(message)
            return
        
        self.sender.reply("❓ 无效的输入,请重新操作")
    
    def view_records(self):
        """查看记录"""
        data = self.get_data()
        
        if not data['records']:
            self.sender.reply("📋 暂无体重记录\n\n💡 发送「体重 65.5」开始记录")
            return
        
        # 按日期排序(最新在前)
        sorted_records = sorted(data['records'], key=lambda x: x['date'], reverse=True)
        
        # 显示最近7条
        display_records = sorted_records[:7]
        
        message = f"📊 体重记录 (共{len(data['records'])}条)\n"
        message += "━━━━━━━━━━━━━━━━━━━━━━━━━\n\n"
        
        for i, record in enumerate(display_records):
            year, month, day = record['date'].split('-')
            
            # 计算趋势
            trend = ""
            if i < len(display_records) - 1:
                prev_weight = display_records[i + 1]['weight']
                diff = record['weight'] - prev_weight
                if diff > 0.1:
                    trend = f" ↑ +{diff:.1f}kg"
                elif diff < -0.1:
                    trend = f" ↓ {diff:.1f}kg"
                else:
                    trend = " → 持平"
            
            message += f"🗓️ {int(month)}月{int(day)}日\n"
            message += f"  📊 {record['weight']}kg{trend}\n\n"
        
        # 显示目标信息
        if data.get('target'):
            latest_weight = sorted_records[0]['weight']
            diff = latest_weight - data['target']
            message += "━━━━━━━━━━━━━━━━━━━━━━━━━\n"
            message += f"🎯 目标体重: {data['target']}kg\n"
            if abs(diff) < 0.1:
                message += "✅ 已达成目标!"
            elif diff > 0:
                message += f"📊 还需减重: {diff:.1f}kg"
            else:
                message += f"📊 还需增重: {abs(diff):.1f}kg"
        
        self.sender.reply(message)
    
    def show_statistics(self):
        """显示统计信息"""
        data = self.get_data()
        
        if not data['records']:
            self.sender.reply("📋 暂无体重记录")
            return
        
        sorted_records = sorted(data['records'], key=lambda x: x['date'])
        
        weights = [r['weight'] for r in sorted_records]
        max_weight = max(weights)
        min_weight = min(weights)
        avg_weight = sum(weights) / len(weights)
        total_change = sorted_records[-1]['weight'] - sorted_records[0]['weight']
        
        max_record = next(r for r in sorted_records if r['weight'] == max_weight)
        min_record = next(r for r in sorted_records if r['weight'] == min_weight)
        
        message = "📊 体重统计\n━━━━━━━━━━━━━━━━━━━━━━━━━\n\n"
        message += f"📈 最高体重: {max_weight}kg\n"
        message += f"   🗓️ {max_record['date']}\n\n"
        message += f"📉 最低体重: {min_weight}kg\n"
        message += f"   🗓️ {min_record['date']}\n\n"
        message += f"📊 平均体重: {avg_weight:.1f}kg\n\n"
        message += "📊 总体变化: "
        if total_change > 0.1:
            message += f"↑ +{total_change:.1f}kg"
        elif total_change < -0.1:
            message += f"↓ {total_change:.1f}kg"
        else:
            message += "→ 基本持平"
        message += f"\n   从 {sorted_records[0]['date']} 到 {sorted_records[-1]['date']}"
        
        if data.get('target'):
            message += "\n\n━━━━━━━━━━━━━━━━━━━━━━━━━\n"
            message += f"🎯 目标体重: {data['target']}kg\n"
            latest_weight = sorted_records[-1]['weight']
            diff = latest_weight - data['target']
            if abs(diff) < 0.1:
                message += "✅ 已达成目标!"
            elif diff > 0:
                message += f"📊 还需减重: {diff:.1f}kg"
            else:
                message += f"📊 还需增重: {abs(diff):.1f}kg"
        
        self.sender.reply(message)
    
    def set_target(self, target_str):
        """设置目标体重"""
        try:
            target = float(target_str)
            if target <= 0 or target > 500:
                self.sender.reply("❌ 目标体重数值无效,请输入0-500之间的数字")
                return
        except ValueError:
            self.sender.reply("❌ 目标体重数值无效,请输入0-500之间的数字")
            return
        
        # 请求确认
        prompt = f"🎯 确认要设置目标体重为: {target}kg 吗？"
        confirmation = self.get_user_confirmation(prompt)
        
        if not confirmation:
            return
        
        if confirmation == "q":
            self.sender.reply("👋 已退出设置流程")
            return
        
        if confirmation == "n":
            self.sender.reply("❌ 已取消设置")
            return
        
        if confirmation == "y":
            data = self.get_data()
            data['target'] = target
            self.save_data(data)
            
            message = f"✅ 已设置目标体重为: {target}kg"
            
            if data['records']:
                sorted_records = sorted(data['records'], key=lambda x: x['date'], reverse=True)
                latest_weight = sorted_records[0]['weight']
                diff = latest_weight - target
                
                message += f"\n\n📊 当前体重: {latest_weight}kg\n"
                if abs(diff) < 0.1:
                    message += "🎉 恭喜!已达成目标!"
                elif diff > 0:
                    message += f"📊 还需减重: {diff:.1f}kg"
                else:
                    message += f"📊 还需增重: {abs(diff):.1f}kg"
            
            self.sender.reply(message)
            return
        
        self.sender.reply("❓ 无效的输入,请重新操作")
    
    def show_target_progress(self):
        """显示目标进度"""
        data = self.get_data()
        
        if not data.get('target'):
            self.sender.reply("❌ 尚未设置目标体重\n\n💡 发送「设置目标体重 60」来设定目标")
            return
        
        if not data['records']:
            self.sender.reply(f"🎯 目标体重: {data['target']}kg\n\n📋 暂无体重记录,无法计算进度")
            return
        
        sorted_records = sorted(data['records'], key=lambda x: x['date'], reverse=True)
        latest_weight = sorted_records[0]['weight']
        diff = latest_weight - data['target']
        
        message = "🎯 目标进度\n━━━━━━━━━━━━━━━━━━━━━━━━━\n\n"
        message += f"📊 当前体重: {latest_weight}kg\n"
        message += f"🎯 目标体重: {data['target']}kg\n\n"
        
        if abs(diff) < 0.1:
            message += "🎉 恭喜!已达成目标体重!\n\n继续保持健康的生活方式!"
        elif diff > 0:
            message += f"📊 还需减重: {diff:.1f}kg\n"
            progress = (1 - diff / latest_weight) * 100
            message += f"📈 进度: {progress:.1f}%"
        else:
            message += f"📊 还需增重: {abs(diff):.1f}kg\n"
            progress = (1 - abs(diff) / data['target']) * 100
            message += f"📈 进度: {progress:.1f}%"
        
        self.sender.reply(message)
    
    def show_help(self):
        """显示帮助信息"""
        help_text = f"📖 体重记录插件 {VERSION}\n\n"
        help_text += "🔹 功能说明：\n"
        help_text += "本插件帮助您记录和追踪体重变化\n\n"
        help_text += "🔹 命令列表：\n"
        help_text += "• 体重 65.5 - 记录当前体重\n"
        help_text += "• 体重记录 - 查看最近记录\n"
        help_text += "• 体重统计 - 查看统计信息\n"
        help_text += "• 设置目标体重 60 - 设定目标体重\n"
        help_text += "• 目标进度 - 查看目标进度\n"
        help_text += "• 体重帮助 - 显示此帮助信息\n\n"
        help_text += "🔹 确认机制：\n"
        help_text += "记录和设置操作需要确认：\n"
        help_text += "  y - 确认执行\n"
        help_text += "  n - 取消操作\n"
        help_text += "  q - 退出流程\n\n"
        help_text += "💡 提示：记录会包含时间信息，方便您追踪健康状况"
        
        self.sender.reply(help_text)
    
    def run(self):
        """主程序入口"""
        try:
            # 路由到对应功能
            if self.message == "体重帮助":
                self.show_help()
            elif self.message == "体重记录":
                self.view_records()
            elif self.message == "体重统计":
                self.show_statistics()
            elif self.message == "目标进度":
                self.show_target_progress()
            elif "设置目标体重" in self.message or "设定目标体重" in self.message:
                match = re.search(r'(?:设置|设定)目标体重\s*([\d.]+)', self.message)
                if match:
                    self.set_target(match.group(1))
                else:
                    self.sender.reply("❓ 请输入目标体重数值 (如: 设置目标体重 60)")
            elif self.message.startswith("体重"):
                # 匹配 "体重 65.5" 格式
                match = re.search(r'体重\s*([\d.]+)', self.message)
                if match:
                    self.record_weight(match.group(1))
                else:
                    self.sender.reply("❓ 未识别的命令\n\n💡 发送「体重帮助」查看使用说明")
            else:
                self.sender.reply("❓ 未识别的命令\n\n💡 发送「体重帮助」查看使用说明")
        
        except Exception as e:
            self.sender.reply(f"❌ 插件执行错误：{e}")


if __name__ == '__main__':
    # 创建插件实例并运行
    plugin = WeightPlugin()
    plugin.run()
